crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,243 @@
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use futures::future::join_all;
+use ipnetwork::IpNetwork;
+use pnet_packet::{
+    ipv4::{checksum, MutableIpv4Packet},
+    udp::MutableUdpPacket,
+};
+use tokio::{net::UdpSocket, sync::RwLock};
+
+use telio_sockets::SocketPool;
+use telio_task::{io::Chan, task_exec, Runtime, RuntimeExt, Task, WaitResponse};
+use telio_utils::{telio_log_debug, telio_log_error, telio_log_warn};
+
+#[cfg(feature = "test-util")]
+const MULTICAST_TRANSPORT_PORT: u16 = 0;
+#[cfg(not(feature = "test-util"))]
+const MULTICAST_TRANSPORT_PORT: u16 = 13569;
+const MAX_PACKET_SIZE: usize = u16::MAX as usize;
+const IPV4_HEADER: usize = 20; // bytes
+const UDP_DEST_OFFSET: usize = 2;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Could not bind to socket")]
+    SocketBindError(String),
+    /// Task execution failed
+    #[error(transparent)]
+    Task(#[from] telio_task::ExecError),
+}
+
+pub struct Config {
+    peers: Vec<IpAddr>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+pub struct Transport {
+    task: Task<State>,
+}
+
+impl Transport {
+    pub async fn start(
+        vpeer_ip: IpAddr,
+        meshnet_ip: IpAddr,
+        socket_pool: Arc<SocketPool>,
+        packet_chan: Chan<Vec<u8>>,
+    ) -> Result<Self, Error> {
+        let transport_socket = socket_pool
+            .new_internal_udp(SocketAddr::new(meshnet_ip, MULTICAST_TRANSPORT_PORT), None)
+            .await
+            .map_err(|e| {
+                telio_log_error!("Failed to bind multicast socket: {:?}", e);
+                Error::SocketBindError(e.to_string())
+            })?;
+        let transport_socket = Arc::new(transport_socket);
+
+        Ok(Self {
+            task: Task::start(State {
+                vpeer_ip,
+                transport_socket,
+                packet_chan,
+                recv_buffer: vec![0; MAX_PACKET_SIZE],
+                nat_map: RwLock::new(HashMap::new()),
+                peers: Vec::new(),
+                multicast_ips: Vec::new(),
+            }),
+        })
+    }
+
+    pub async fn configure(&self, config: Config) -> Result<(), Error> {
+        task_exec!(&self.task, async move |state| {
+            state.configure(config);
+            Ok(())
+        })
+        .await?;
+        Ok(())
+    }
+
+    pub async fn stop(self) {
+        let _ = self.task.stop().await.resume_unwind();
+    }
+}
+
+struct State {
+    vpeer_ip: IpAddr,
+    transport_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat_map: RwLock<HashMap<u16, SocketAddr>>,
+    peers: Vec<IpAddr>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+impl State {
+    fn configure(&mut self, config: Config) {
+        self.peers = config.peers;
+        self.multicast_ips = config.multicast_ips;
+    }
+
+    async fn handle_local_multicast_packet(&mut self, packet: Vec<u8>) {
+        let fanout_res = join_all(self.peers.iter().map(|peer| {
```

Comment:
This is one area, there it would be nice to have pk's also, and to have concrete peers we failed to send to.

Reply:
in trace bellow